src/mailbox.cpp

```diff
@@ -11,64 +11,78 @@
 
 extern Game g_game;
 
-ReturnValue Mailbox::queryAdd(int32_t, const Thing& thing, uint32_t, uint32_t, Creature*) const
+static bool canSend(std::shared_ptr<const Item> item)
 {
-	const Item* item = thing.getItem();
-	if (item && Mailbox::canSend(item)) {
+	return item->getID() == ITEM_PARCEL || item->getID() == ITEM_LETTER;
+}
+
+ReturnValue Mailbox::queryAdd(int32_t, std::shared_ptr<const Thing> thing, uint32_t, uint32_t,
+                              std::shared_ptr<Creature>) const
+{
+	auto item = thing->getItem();
+	if (item && canSend(item)) {
 		return RETURNVALUE_NOERROR;
 	}
 	return RETURNVALUE_NOTPOSSIBLE;
 }
 
-ReturnValue Mailbox::queryMaxCount(int32_t, const Thing&, uint32_t count, uint32_t& maxQueryCount, uint32_t) const
+ReturnValue Mailbox::queryMaxCount(int32_t, std::shared_ptr<const Thing>, uint32_t count, uint32_t& maxQueryCount,
+                                   uint32_t) const
 {
 	maxQueryCount = std::max<uint32_t>(1, count);
 	return RETURNVALUE_NOERROR;
 }
 
-ReturnValue Mailbox::queryRemove(const Thing&, uint32_t, uint32_t, Creature* /*= nullptr */) const
+ReturnValue Mailbox::queryRemove(std::shared_ptr<const Thing>, uint32_t, uint32_t,
+                                 std::shared_ptr<Creature> /*= nullptr */) const
 {
 	return RETURNVALUE_NOTPOSSIBLE;
 }
 
-Cylinder* Mailbox::queryDestination(int32_t&, const Thing&, Item**, uint32_t&) { return this; }
+std::shared_ptr<Thing> Mailbox::queryDestination(int32_t&, std::shared_ptr<const Thing>, std::shared_ptr<Item>&,
+                                                 uint32_t&)
+{
+	return shared_from_this();
+}
 
-void Mailbox::addThing(Thing* thing) { return addThing(0, thing); }
+void Mailbox::addThing(std::shared_ptr<Thing> thing) { return addThing(0, thing); }
 
-void Mailbox::addThing(int32_t, Thing* thing)
+void Mailbox::addThing(int32_t, std::shared_ptr<Thing> thing)
 {
-	Item* item = thing->getItem();
-	if (item && Mailbox::canSend(item)) {
+	auto item = thing->getItem();
```

Comment:
Could this be `const auto`
```suggestion
	const auto item = thing->getItem();
```